File: app/src/main/java/com/cardee/owner_profile_info/presenter/OwnerProfileInfoPresenter.java
package com.cardee.owner_profile_info.presenter;

import android.content.Context;
import android.content.DialogInterface;
import android.location.Address;
import android.location.Geocoder;
import android.util.Log;

import com.cardee.R;
import com.cardee.data_source.Error;
import com.cardee.data_source.remote.api.profile.request.OwnerNoteRequest;
import com.cardee.data_source.remote.api.profile.response.entity.OwnerProfile;
import com.cardee.data_source.remote.api.profile.response.entity.OwnerReview;
import com.cardee.data_source.util.DialogHelper;
import com.cardee.domain.UseCase;
import com.cardee.domain.UseCaseExecutor;
import com.cardee.domain.owner.entity.Car;
import com.cardee.domain.owner.entity.CarReview;
import com.cardee.domain.owner.entity.mapper.OwnerReviewToCarReviewMapper;
import com.cardee.domain.owner.usecase.ChangeImage;
import com.cardee.domain.owner.usecase.ChangeNote;
import com.cardee.domain.owner.usecase.GetOwnerInfo;
import com.cardee.owner_profile_info.view.ProfileInfoView;

import java.io.File;
import java.io.IOException;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;

import io.reactivex.functions.Consumer;

public class OwnerProfileInfoPresenter implements Consumer<Car> {

    private final String TAG = this.getClass().getSimpleName();

    private final GetOwnerInfo mGetInfoUseCase;
    private final ChangeNote mChangeNote;
    private final ChangeImage mChangeImage;
    private UseCaseExecutor mExecutor;
    private ProfileInfoView mView;

    public OwnerProfileInfoPresenter(ProfileInfoView view) {
        mGetInfoUseCase = new GetOwnerInfo();
        mChangeNote = new ChangeNote();
        mChangeImage = new ChangeImage();
        mExecutor = UseCaseExecutor.getInstance();
        mView = view;
    }

    public void getOwnerInfo() {
        mView.showProgress(true);

        mExecutor.execute(mGetInfoUseCase, null, new UseCase.Callback<GetOwnerInfo.ResponseValues>() {
            @Override
            public void onSuccess(GetOwnerInfo.ResponseValues response) {
                if (mView != null) {
                    OwnerProfile profile = response.getOwnerProfile();
                    if (profile != null) {
                        mView.setProfileImage(profile.getProfilePhotoLink());
                        mView.setProfileName(profile.getName());
                        float rate = profile.getRating();
                        if (rate == (long) rate) {
                            mView.setProfileRating(String.format(Locale.getDefault(), "%d", (long) rate));
                        } else {
                            mView.setProfileRating(String.format(Locale.getDefault(), "%.1f", rate));
                        }

                        String acceptance = profile.getAcceptance();
                        if (acceptance == null) {
                            acceptance = "0";
                        }
                        mView.setAcceptance(acceptance);

                        String responseTime = profile.getResponseTime();
                        if (responseTime == null) {
                            responseTime = "0";
                        }
                        mView.setResponseText(responseTime);

                        String minutes = ((Context) mView).getResources().getString(R.string.owner_profile_info_minutes);
                        if (Integer.valueOf(responseTime) != 1) {
                            minutes = minutes + "s";
                        }
                        mView.setMinutes(minutes);

                        mView.setBookings(profile.getBookingsCount().toString());

                        mView.setNote(profile.getNote());

                        String address = profile.getAddress();
                        if (address != null && !address.isEmpty()) {
                            Geocoder geocoder = new Geocoder((Context) mView, Locale.getDefault());
                            try {
                                List<Address> addresses = geocoder.getFromLocationName(address, 1);
                                String city = addresses.get(0).getLocality();
                                if (city == null) {
                                    city = ((Context) mView).getResources().getString(R.string.owner_profile_info_default_city);
                                }
                                mView.setNoteTitle(city);
                            } catch (IOException e) {
                                Log.d(TAG, e.getMessage());
                            }
                        }


                        int carsCount = profile.getCarCount();
                        StringBuilder builder = new StringBuilder(String.valueOf(carsCount));
                        builder.append(((Context) mView).getResources().getString(R.string.owner_profile_info_car));
                        if (carsCount != 1) {
                            builder.append("s");
                        }
                        mView.setCarsCount(builder.toString());

                        int reviewsCount = profile.getReviewCount();
                        builder = new StringBuilder(String.valueOf(reviewsCount));
                        builder.append(((Context) mView).getResources().getString(R.string.owner_profile_info_car_review));
                        if (reviewsCount != 1) {
                            builder.append("s");
                        }

                        mView.setReviewsCount(builder.toString());
                    }

                    List<Car> cars = response.getCars();
                    if (cars != null && !cars.isEmpty()) {
                        mView.setCars(cars);
                    }

                    OwnerReview[] reviews = profile.getReviews();
                    if (reviews != null && reviews.length > 0) {
                        OwnerReviewToCarReviewMapper mapper = new OwnerReviewToCarReviewMapper();
                        List<CarReview> carReviewList = mapper.transform(reviews);
                        Iterator<CarReview> iterator = carReviewList.iterator();
                        while (iterator.hasNext()) {
                            String text = iterator.next().getReviewText();
                            if (text == null || text.isEmpty()) {
                                iterator.remove();
                            }
                        }
                        mView.setCarReviews(carReviewList);
                    }

                    mView.showProgress(false);
                }
            }

            @Override
            public void onError(Error error) {
                mView.showProgress(false);
                mView.showMessage(error.getMessage());
            }
        });
    }

    @Override
    public void accept(Car car) throws Exception {
        mView.openItem(car);
    }

    public void changeNote(final Context context) {
        DialogHelper.getAlertDialog(context, R.layout.dialog_owner_profile_change_note,
                context.getResources().getString(R.string.owner_profile_info_note_change),
                new DialogHelper.OnClickCallback() {
                    @Override
                    public void onPositiveButtonClick(final String newNote, final DialogInterface dialog) {
                        OwnerNoteRequest ownerNoteRequest = new OwnerNoteRequest();
                        ownerNoteRequest.setNote(newNote);
                        mExecutor.execute(mChangeNote, new ChangeNote.RequestValues(ownerNoteRequest), new UseCase.Callback<ChangeNote.ResponseValues>() {
                            @Override
                            public void onSuccess(ChangeNote.ResponseValues response) {
                                dialog.dismiss();
                                mView.setNote(newNote);
                                mView.showMessage(R.string.owner_profile_info_note_change_success);
                            }

                            @Override
                            public void onError(Error error) {
                                mView.showMessage(error.getMessage());
                            }
                        });
                    }

                    @Override
                    public void onNegativeButtonClick(DialogInterface dialog) {
                        dialog.dismiss();
                    }
                }).show();
    }

    public void setProfilePicture(File photo) {
        mView.showProgress(true);
        mExecutor.execute(mChangeImage,
                new ChangeImage.RequestValues(null, null, photo, null),
                new UseCase.Callback<ChangeImage.ResponseValues>() {
                    @Override
                    public void onSuccess(ChangeImage.ResponseValues response) {
                        mView.showProgress(false);
                        mView.showMessage(R.string.owner_profile_info_photo_success);
                        mView.onChangeImageSuccess();
                    }

                    @Override
                    public void onError(Error error) {
                        mView.showProgress(false);
                        mView.showMessage(error.getMessage());
                    }
                });
    }

}

File: app/src/main/java/com/cardee/owner_profile_info/presenter/OwnerProfileInfoPresenter.kt
package com.cardee.owner_profile_info.presenter

import android.content.Context
import android.content.DialogInterface
import android.content.res.Resources
import android.location.Geocoder
import com.cardee.CardeeApp
import com.cardee.R
import com.cardee.data_source.Error
import com.cardee.data_source.remote.api.profile.request.ChangeNoteRequest
import com.cardee.data_source.remote.api.profile.response.entity.OwnerProfile
import com.cardee.data_source.util.DialogHelper
import com.cardee.domain.UseCase
import com.cardee.domain.UseCaseExecutor
import com.cardee.domain.owner.entity.Car
import com.cardee.domain.owner.entity.mapper.OwnerReviewToCarReviewMapper
import com.cardee.domain.owner.usecase.ChangeImage
import com.cardee.domain.owner.usecase.ChangeNote
import com.cardee.domain.owner.usecase.GetOwnerInfo
import com.cardee.domain.owner.usecase.GetOwnerInfoById
import com.cardee.owner_profile_info.view.ProfileInfoView
import io.reactivex.functions.Consumer
import java.io.File
import java.util.*

class OwnerProfileInfoPresenter(val mView: ProfileInfoView?) : Consumer<Car> {

    private val mGetInfoUseCase: GetOwnerInfo
    private val mGetInfoById: GetOwnerInfoById
    private val mChangeNote: ChangeNote
    private val mChangeImage: ChangeImage
    private val mExecutor: UseCaseExecutor
    private val mResources: Resources

    init {
        mGetInfoUseCase = GetOwnerInfo()
        mGetInfoById = GetOwnerInfoById()
        mChangeNote = ChangeNote()
        mChangeImage = ChangeImage()
        mExecutor = UseCaseExecutor.getInstance()
        mResources = (mView as Context).resources
    }

    fun getOwnerById(profileId: Int) {
        mView?.showProgress(true)
        mExecutor.execute(mGetInfoById, GetOwnerInfoById.RequestValues(profileId), object : UseCase.Callback<GetOwnerInfoById.ResponseValues> {
            override fun onSuccess(response: GetOwnerInfoById.ResponseValues?) {
                setProfileFields(response)
                mView?.showProgress(false)
            }

            override fun onError(error: Error?) {
                mView?.showProgress(false)
                mView?.showMessage(error?.message)
            }
        })
    }

    fun getOwnerInfo() {
        mView?.showProgress(true)
        mExecutor.execute(mGetInfoUseCase, null, object : UseCase.Callback<GetOwnerInfo.ResponseValues> {
            override fun onSuccess(response: GetOwnerInfo.ResponseValues?) {
                mView?.showProgress(false)
                setProfileFields(response)
            }

            override fun onError(error: Error?) {
                mView?.showProgress(false)
                mView?.showMessage(error?.message)
            }
        })
    }

    private fun setProfileFields(response: GetOwnerInfo.ResponseValues?) {
        val profile = response?.ownerProfile
        if (profile != null) {
            mView?.setProfileImage(profile.profilePhotoLink)
            mView?.setProfileName(profile.name)
            setProfileRating(profile)
            setProfileAcceptance(profile)
            setResponseTime(profile)
            mView?.setBookings(profile.bookingsCount.toString())
            setNote(profile)
            setCarsCount(profile)
            setReviewsCount(profile)
            setCars(response)
            setReviews(profile)
        }
    }

    private fun setReviews(profile: OwnerProfile) {
        val reviews = profile.reviews
        if (reviews.isNotEmpty()) {
            val mapper = OwnerReviewToCarReviewMapper()
            val carReviewList = mapper.transform(reviews)
            val iterator = carReviewList.iterator()
            while (iterator.hasNext()) {
                val review = iterator.next()
                val text = review.reviewText
                if (text.isNullOrEmpty()) {
                    iterator.remove()
                } else {
                    review.reviewText = text.trim()
                }
            }
            carReviewList.sortByDescending { review -> review.reviewDate }
            mView?.setCarReviews(carReviewList)
        }
    }

    private fun setCars(response: GetOwnerInfo.ResponseValues) {
        val cars: MutableList<Car>? = response.cars
        if (cars?.isNotEmpty() == true) {
            mView?.setCars(cars)
        }
    }

    private fun setReviewsCount(profile: OwnerProfile) {
        val reviewsCount = profile.reviewCount
        val builder = StringBuilder(reviewsCount.toString())
        builder.append(mResources.getString(R.string.owner_profile_info_car_review))
        if (reviewsCount != 1) {
            builder.append("s")
        }
        mView?.setReviewsCount(builder.toString())
    }

    private fun setCarsCount(profile: OwnerProfile) {
        val carsCount = profile.carCount
        val builder = StringBuilder(carsCount.toString())
        builder.append(mResources.getString(R.string.owner_profile_info_car))
        if (carsCount != 1) {
            builder.append("s")
        }
        mView?.setCarsCount(builder.toString())
    }

    private fun setNote(profile: OwnerProfile) {
        var note = profile.note
        if (note.isNullOrEmpty()) {
            note = mResources.getString(R.string.profile_default_note)
        }
        mView?.setNote(note)

        var city: String? = ""
        val address = profile.address
        if (!address.isNullOrEmpty()) {
            val geocoder = Geocoder(CardeeApp.context, Locale.getDefault())
            var addressList = geocoder.getFromLocationName(address, 1)
            if (addressList.isNotEmpty()) {
                city = addressList[0].locality
            }
        }
        if (city.isNullOrEmpty()) {
            city = mResources.getString(R.string.owner_profile_info_default_city)
        }
        mView?.setNoteTitle(city)
    }

    private fun setResponseTime(profile: OwnerProfile) {
        val responseTime = profile.responseTime
        if (responseTime == null) {
            mView?.setResponseText("0")
        } else {
            mView?.setResponseText(responseTime)
        }

        var minutes = mResources.getString(R.string.owner_profile_info_minute)
        if (responseTime.toInt() != 1) {
            minutes += "s"
        }
        mView?.setMinutes(minutes)
    }

    private fun setProfileAcceptance(profile: OwnerProfile) {
        val accept: Float? = profile.acceptance
        if (accept == null) {
            mView?.setAcceptance("0")
        } else {
            if (accept % 1 == 0F) {
                mView?.setAcceptance(String.format(Locale.getDefault(), "%d", accept.toInt()))
            } else {
                mView?.setAcceptance(String.format(Locale.getDefault(), "%.1f", accept))
            }
        }
    }

    private fun setProfileRating(profile: OwnerProfile) {
        val rate: Float? = profile.rating
        if (rate != null) {
            if (rate % 1 == 0F) {
                mView?.setProfileRating(String.format(Locale.getDefault(), "%d", rate.toInt()))
            } else {
                mView?.setProfileRating(String.format(Locale.getDefault(), "%.1f", rate))
            }
        }
    }

    override fun accept(car: Car?) {
        mView?.openItem(car)
    }

    fun changeNote(context: Context) {
        DialogHelper.getAlertDialog(context, R.layout.dialog_profile_change_note,
                mResources.getString(R.string.profile_info_note_change_title),
                mResources.getString(R.string.profile_info_note_change), object : DialogHelper.OnClickCallback {
            override fun onPositiveButtonClick(newNote: String?, dialog: DialogInterface?) {
                val changeNoteRequest = ChangeNoteRequest()
                changeNoteRequest.note = newNote
                mExecutor.execute(mChangeNote, ChangeNote.RequestValues(changeNoteRequest), object : UseCase.Callback<ChangeNote.ResponseValues> {
                    override fun onSuccess(response: ChangeNote.ResponseValues?) {
                        dialog?.dismiss()
                        mView?.setNote(newNote)
                        mView?.showMessage(R.string.profile_info_note_change_success)
                    }

                    override fun onError(error: Error?) {
                        dialog?.dismiss()
                        mView?.showMessage(error?.message)
                    }
                })
            }

            override fun onNegativeButtonClick(dialog: DialogInterface?) {
                dialog?.dismiss()
            }
        }).show()
    }

    fun setProfilePicture(photo: File) {
        mView?.showProgress(true)
        mExecutor.execute(mChangeImage, ChangeImage.RequestValues(null, null, photo, null), object : UseCase.Callback<ChangeImage.ResponseValues> {
            override fun onSuccess(response: ChangeImage.ResponseValues?) {
                mView?.showProgress(false)
                mView?.showMessage(R.string.owner_profile_info_photo_success)
                mView?.onChangeImageSuccess()
            }

            override fun onError(error: Error?) {
                mView?.showProgress(false)
                mView?.showMessage(error?.message)
            }
        })
    }
}

